src/swap-auto-generator/swap-auto-generator.c

```diff
@@ -0,0 +1,436 @@
+        struct udev_device *parent = NULL;
+        const char *name, *node, *pttype, *devtype;
+        blkid_partlist pl;
+        int r;
+        dev_t pn;
+
+        udev = udev_new();
+        if (!udev)
+                return log_oom();
+
+        d = udev_device_new_from_devnum(udev, 'b', devnum);
+        if (!d)
+                return log_oom();
+
+        name = udev_device_get_devnode(d);
+        if (!name)
+                name = udev_device_get_syspath(d);
+        if (!name) {
+                log_debug("Device %u:%u does not have a name, ignoring.",
+                          major(devnum), minor(devnum));
+                return 0;
+        }
+
+        parent = udev_device_get_parent(d);
+        if (!parent) {
+                log_debug("%s: not a partitioned device, ignoring.", name);
+                return 0;
+        }
+
+        /* Does it have a devtype? */
+        devtype = udev_device_get_devtype(parent);
+        if (!devtype) {
+                log_debug("%s: parent doesn't have a device type, ignoring.", name);
+                return 0;
+        }
+
+        /* Is this a disk or a partition? We only care for disks... */
+        if (!streq(devtype, "disk")) {
+                log_debug("%s: parent isn't a raw disk, ignoring.", name);
+                return 0;
+        }
+
+        /* Does it have a device node? */
+        node = udev_device_get_devnode(parent);
+        if (!node) {
+                log_debug("%s: parent device does not have device node, ignoring.", name);
+                return 0;
+        }
+
+        log_debug("%s: root device %s.", name, node);
+
+        pn = udev_device_get_devnum(parent);
+        if (major(pn) == 0)
+                return 0;
+
+        errno = 0;
+        b = blkid_new_probe_from_filename(node);
+        if (!b) {
+                if (errno == 0)
+                        return log_oom();
+
+                return log_error_errno(errno, "%s: failed to allocate prober: %m", node);
+        }
+
+        blkid_probe_enable_partitions(b, 1);
+        blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
+
+        errno = 0;
+        r = blkid_do_safeprobe(b);
+        if (r == 1)
+                return 0; /* no results */
+        else if (r == -2) {
+                log_warning("%s: probe gave ambiguous results, ignoring.", node);
+                return 0;
+        } else if (r != 0)
+                return log_error_errno(errno ?: EIO, "%s: failed to probe: %m", node);
+
+        errno = 0;
+        r = blkid_probe_lookup_value(b, "PTTYPE", &pttype, NULL);
+        if (r != 0) {
+                if (errno == 0)
+                        return 0; /* No partition table found. */
+
+                return log_error_errno(errno, "%s: failed to determine partition table type: %m", node);
+        }
+
+        /* We only do this all for MSDOS */
+        log_debug("partition table type: %s.", pttype);
+        //if (!streq_ptr(pttype, "dos")) {
```

Comment:
oh, sorry, I meant to update it once I confirmed the correct string for the partition table type. I'll fix this.